README.md

# RClimacell <a href='https://nikdata.github.io/RClimacell/'><img src='man/figures/rclimacell-hex.png' align="right" width="150" height="150" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/figures/lifecycle-experimental.svg)
[![License:
MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![GitHub
commit](https://img.shields.io/github/last-commit/nikdata/RClimacell)](https://github.com/nikdata/RClimacell/commit/main)
[![R-CMD-check](https://github.com/nikdata/RClimacell/workflows/R-CMD-check/badge.svg)](https://github.com/nikdata/RClimacell/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/RClimacell)](https://CRAN.R-project.org/package=RClimacell)
<!-- badges: end -->

The {RClimacell} package is an **unofficial** R package that enables
basic interaction with [Climacell’s](https://www.climacell.co) API using
the [Timeline
Interface](https://docs.climacell.co/reference/timeline-overview). The
functions within this package are tested against some of the [Core data
layers](https://docs.climacell.co/reference/data-layers-core).

Please note that using the functions within this package **require** a
valid API key.

More information about the Climacell API can be found on their
[docs](https://docs.climacell.co/reference/api-overview) page.

### Lubridate Issue

As of 24 Feb, there is a [known
issue](https://github.com/tidyverse/lubridate/issues/928) with using the
package {lubridate} and it seems to be affecting macOS users. The ‘fix’
has been to add the following line to the .Renviron file or the
.Rprofile (I applied the code into the .Renviron file and it worked):

``` r
TZDIR="/Library/Frameworks/R.framework/Resources/share/zoneinfo/"
```

{lubridate} version 1.7.10 fixes this issue and is available on CRAN.

## Installation

CRAN version can be installed as follows:

``` r
install.packages('RClimacell')
```

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("nikdata/RClimacell")
```

## Usage

Not every variable in each of the functions will have a value. Missing
values are denoted by NA and indicate that the API did not return a
value for the specific date/time and function call.

### Temperature

``` r
library(RClimacell)
climacell_temperature(api_key = Sys.getenv("CLIMACELL_API"),
                      lat = 41.878685,
                      long = -87.636011,
                      timestep = '1d',
                      start_time = lubridate::now(),
                      end_time = lubridate::now() + lubridate::days(3))
#> # A tibble: 4 x 5
#>   start_time          temp_c temp_feel_c dewpoint humidity
#>   <dttm>               <dbl>       <dbl>    <dbl>    <dbl>
#> 1 2021-03-22 11:00:00  18.3        18.2      5.88     62.1
#> 2 2021-03-23 11:00:00  14.1        14.1     10.5      97.0
#> 3 2021-03-24 11:00:00  14.2        14.2     10.1      97.3
#> 4 2021-03-25 11:00:00   7.26        7.26     4.94     92.3
```

### Wind

``` r
library(RClimacell)
climacell_wind(api_key = Sys.getenv("CLIMACELL_API"),
               lat = 41.878685,
               long = -87.636011,
               timestep = '1d',
               start_time = lubridate::now(),
               end_time = lubridate::now() + lubridate::days(3))
#> # A tibble: 4 x 4
#>   start_time          wind_speed wind_gust wind_direction
#>   <dttm>                   <dbl>     <dbl>          <dbl>
#> 1 2021-03-22 11:00:00       6.46      9.12          190. 
#> 2 2021-03-23 11:00:00       9.77     13.9           143. 
#> 3 2021-03-24 11:00:00      10.7      15.4           225. 
#> 4 2021-03-25 11:00:00       6.77      9.31           95.1
```

### Precipitation

``` r
library(RClimacell)
df_precip <- climacell_precip(api_key = Sys.getenv("CLIMACELL_API"),
                 lat = 41.878685,
                 long = -87.636011,
                 timestep = '1h',
                 start_time = lubridate::now(),
                 end_time = lubridate::now() + lubridate::days(3))

dplyr::glimpse(df_precip)
#> Rows: 73
#> Columns: 14
#> $ start_time                <dttm> 2021-03-22 22:00:00, 2021-03-22 23:00:00, 2…
#> $ precipitation_intensity   <dbl> 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.00…
#> $ precipitation_probability <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ precipitation_type_code   <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ precipitation_type_desc   <chr> "Rain", "Rain", "Rain", "Rain", "Rain", "Rai…
#> $ visibility                <dbl> 10.00, 16.00, 16.00, 16.00, 16.00, 16.00, 16…
#> $ pressure_surface_level    <dbl> 997.45, 996.97, 997.25, 997.07, 997.85, 998.…
#> $ pressure_sea_level        <dbl> 1015.03, 1015.30, 1015.48, 1015.25, 1016.09,…
#> $ cloud_cover               <dbl> 21.43, 82.86, 100.00, 100.00, 100.00, 100.00…
#> $ cloud_base                <dbl> NA, 7.36, 1.03, 0.74, 0.57, 7.77, 6.67, 6.67…
#> $ cloud_ceiling             <dbl> NA, NA, 9.58, 9.19, 8.64, 7.64, 6.66, 6.53, …
#> $ solar_ghi                 <dbl> 247.85, 139.36, 0.00, 0.00, 0.00, 0.00, 0.00…
#> $ weather_code              <dbl> 1100, 1001, 1001, 1001, 1001, 1001, 1001, 10…
#> $ weather_desc              <chr> "Mostly Clear", "Cloudy", "Cloudy", "Cloudy"…
```

### Celestial (sunset time, sunrise time, and moon phase)

``` r
library(RClimacell)
df_celestial <- climacell_celestial(api_key = Sys.getenv("CLIMACELL_API"),
                 lat = 41.878685,
                 long = -87.636011,
                 timestep = '1d',
                 start_time = lubridate::now(),
                 end_time = lubridate::now() + lubridate::days(5))

dplyr::glimpse(df_celestial)
#> Rows: 6
#> Columns: 5
#> $ start_time             <dttm> 2021-03-22 11:00:00, 2021-03-23 11:00:00, 2021…
#> $ sunrise_time           <dttm> 2021-03-22 11:50:00, 2021-03-23 11:48:20, 2021…
#> $ sunset_time            <dttm> 2021-03-23 00:05:00, 2021-03-24 00:06:40, 2021…
#> $ moon_phase_code        <int> 2, 2, 2, 2, 3, 4
#> $ moon_phase_description <chr> "First Quarter", "First Quarter", "First Quarte…
```

### Climacell Core (all Core Layer data)

This function aims to retrieve all of the Core Layer data using the
Timeline Interface. All of the data are retrieved in a single API call.
Note that if the timestep is not ‘1d’, then the moon phase, sunrise
time, and sunset times will not be available

``` r
library(RClimacell)
df_core <- climacell_core(api_key = Sys.getenv("CLIMACELL_API"),
                 lat = 41.878685,
                 long = -87.636011,
                 timestep = '1m',
                 start_time = lubridate::now(),
                 end_time = lubridate::now() + lubridate::hours(3))
#> Moonphase, Sunrise Time, and Sunset Times are only available if timestep is '1d'.

dplyr::glimpse(df_core)
#> Rows: 181
#> Columns: 21
#> $ start_time                <dttm> 2021-03-22 22:00:00, 2021-03-22 22:01:00, 2…
#> $ temp_c                    <dbl> 16.11, 16.11, 16.11, 16.11, 16.11, 16.11, 16…
#> $ temp_feel_c               <dbl> 17.81, 17.82, 17.83, 17.83, 17.84, 17.85, 17…
#> $ weather_code              <dbl> 1100, 1100, 1100, 1100, 1100, 1100, 1100, 11…
#> $ weather_desc              <chr> "Mostly Clear", "Mostly Clear", "Mostly Clea…
#> $ dewpoint                  <dbl> 4.08, 4.08, 4.08, 4.08, 4.08, 4.08, 4.09, 4.…
#> $ humidity                  <dbl> 45.00, 45.00, 45.00, 45.00, 45.00, 45.00, 45…
#> $ wind_speed                <dbl> 0.89, 0.89, 0.89, 0.89, 0.89, 0.89, 0.89, 0.…
#> $ wind_direction            <dbl> 106, 106, 106, 106, 106, 106, 106, 106, 106,…
#> $ wind_gust                 <dbl> 1.78, 1.78, 1.78, 1.78, 1.78, 1.78, 1.79, 1.…
#> $ solar_ghi                 <dbl> 247.85, 246.04, 244.23, 242.42, 240.62, 238.…
#> $ precipitation_type_code   <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ precipitation_type_desc   <chr> "Rain", "Rain", "Rain", "Rain", "Rain", "Rai…
#> $ precipitation_probability <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ precipitation_intensity   <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ cloud_cover               <dbl> 21.43, 21.43, 21.43, 21.43, 21.43, 21.43, 21…
#> $ cloud_base                <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
#> $ cloud_ceiling             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
#> $ visibility                <dbl> 10.00, 10.00, 10.00, 10.00, 10.00, 10.00, 10…
#> $ pressure_surface_level    <dbl> 997.45, 997.44, 997.43, 997.42, 997.41, 997.…
#> $ pressure_sea_level        <dbl> 1015.03, 1015.03, 1015.03, 1015.03, 1015.03,…
```

``` r
library(RClimacell)
df_core2 <- climacell_core(api_key = Sys.getenv("CLIMACELL_API"),
                 lat = 41.878685,
                 long = -87.636011,
                 timestep = '1d',
                 start_time = lubridate::now(),
                 end_time = lubridate::now() + lubridate::days(5))

dplyr::glimpse(df_core2)
#> Rows: 6
#> Columns: 25
#> $ start_time                <dttm> 2021-03-22 11:00:00, 2021-03-23 11:00:00, 2…
#> $ temp_c                    <dbl> 18.22, 14.07, 14.18, 7.26, 2.57, 3.00
#> $ temp_feel_c               <dbl> 18.22, 14.07, 14.18, 7.26, -0.81, -0.95
#> $ weather_code              <dbl> 1102, 4200, 4000, 5001, 5001, 5100
#> $ weather_desc              <chr> "Mostly Cloudy", "Light Rain", "Drizzle", "F…
#> $ dewpoint                  <dbl> 5.88, 10.49, 10.07, 4.94, -1.81, 2.25
#> $ humidity                  <dbl> 62.11, 96.98, 97.27, 92.32, 74.93, 90.36
#> $ wind_speed                <dbl> 6.46, 9.77, 10.72, 9.75, 7.94, 4.69
#> $ wind_direction            <dbl> 187.73, 142.57, 224.78, 61.08, 148.06, 98.70
#> $ wind_gust                 <dbl> 9.12, 13.92, 15.36, 12.24, 9.84, 5.95
#> $ solar_ghi                 <dbl> 513.56, 540.69, 552.39, 557.51, 651.40, 137.…
#> $ precipitation_type_code   <dbl> 1, 1, 1, 2, 2, 2
#> $ precipitation_type_desc   <chr> "Rain", "Rain", "Rain", "Snow", "Snow", "Sno…
#> $ precipitation_probability <dbl> 0, 75, 25, 55, 20, 30
#> $ precipitation_intensity   <dbl> 0.0000, 2.7397, 0.3605, 0.4792, 0.0222, 0.57…
#> $ cloud_cover               <dbl> 100, 100, 100, 100, 100, 100
#> $ cloud_base                <dbl> 7.77, 2.68, 5.70, 5.86, 1.09, 2.18
#> $ cloud_ceiling             <dbl> 9.58, 8.31, 7.31, 8.26, 1.26, 3.35
#> $ visibility                <dbl> 16.00, 16.00, 16.00, 24.14, 24.14, 24.14
#> $ pressure_surface_level    <dbl> 999.40, 996.42, 990.88, 992.93, 1001.49, 999…
#> $ pressure_sea_level        <dbl> 1014.74, 1003.36, 1003.40, 1011.24, 1015.15,…
#> $ sunrise_time              <dttm> 2021-03-22 11:50:00, 2021-03-23 11:48:20, 20…
#> $ sunset_time               <dttm> 2021-03-23 00:05:00, 2021-03-24 00:06:40, 20…
#> $ moon_phase_code           <dbl> 2, 2, 2, 2, 3, 4
#> $ moon_phase_description    <chr> "First Quarter", "First Quarter", "First Qua…
```

See the [vignette](https://nikdata.github.io/RClimacell/) for more
information.


NEWS.md
# RClimacell 0.1.4

* added new function `climacell_core()` that retrieves all core layer data from Timeline Interface
* added new function `climacell_celestial()` that retrieves the sunrise and sunset times along with the moon phase
* removed the need to use the package {parsedate} and removed {parsedate} from dependency list
* input dates no longer have to be in ISO8601 format (recommended to use the {lubridate} package or `Sys.time()` when entering date-time stamps)
* drastically improved error handling for dates/times provided by user
* updated vignettes

# RClimacell 0.1.3

* fixed a bug where the user provided API key was not used

# RClimacell 0.1.2

* Climacell API does not return results unless non-zero values are present for pressure; added extra steps to account for this.

# RClimacell 0.1.1

* fixed a bug where the `climacell_precip()` function would not return results for the correct lat/long and `timestep` argument.
* fixed a coding error where the start and end times were not parameterized prior to calling the API.
* improved the way cloud cover results are handled when using the `climacell_precip()` function.

# RClimacell 0.1.0

* Initial release
* Added a `NEWS.md` file to track changes to the package.



R/climacell_core.R
#' Climacell Core Layer Data
#'
#' @description \code{climacell_core} returns a tibble that contains all of the Core Layer data from the Climacell version 4 API using the Timelines interface. The intent of this function is to retrieve all of the Core Layer data in a single API call. This is especially handy when using the free API as it limits the usage of the API based on hourly rate and daily usage.
#'
#' @param api_key character string representing the private API key. Provided by user or loaded automatically from environment variable (environment variable must be called "CLIMACELL_API").
#' @param lat a numeric value (or a string that can be coerced to numeric) representing the latitude of the location.
#' @param long a numeric value (or a string that can be coerced to numeric) representing the longitude of the location.
#' @param timestep a 'step' value for the time. Choose one of the following valid values: c('1d', '1h', '30m','15m','5m','1m','current').
#' @param start_time the start time of the query. This input must be a character string that can be parsed into a data/time or a date/time value. If the input does not contain a timezone, the value will be assumed to be in UTC. It is recommended that the \code{lubridate::now()} function or \code{Sys.time()} be used to define the start_time. For this function, the start_time cannot be less than 6 hours from the current time.
#' @param end_time the end time of the query. This input must be a character string that can be parsed into a data/time or a date/time value. If the input does not contain a timezone, the value will be assumed to be in UTC. OPTIONAL if timestep is 'current' or if the user desires to get the maximum results possible (depends on the timestep chosen).
#'
#' @return a tibble
#' @export
#'
#' @import dplyr
#' @import tibble
#' @import httr
#' @import utils
#' @import tidyselect
#'
#' @importFrom stringr str_detect
#' @importFrom magrittr `%>%`
#' @importFrom rlang .data
#' @importFrom tidyr fill pivot_wider
#'
#' @examples
#' \dontrun{
#' climacell_core(
#'   api_key = Sys.getenv('CLIMACELL_API'),
#'   lat = 0,
#'   long = 0,
#'   timestep = '1d',
#'   start_time = lubridate::now(),
#'   end_time = lubridate::now + lubridate::days(5))
#' }
#'

climacell_core <- function(api_key, lat, long, timestep, start_time=NULL, end_time=NULL) {

  # retrieve the appropriate dictionaries
  moon_dict <- dict_moonphase()
  precip_dict <- dict_preciptype()
  weather_dict <- dict_weathercode()

  # check for missing key or empty environment variable
  if(missing(api_key) & Sys.getenv('CLIMACELL_API') == '') {
    stop("No API key provided nor default CLIMACELL_API environment variable found.\nPlease provide valid API key.")
  }

  # basic way of assigning environment variable
  # if no api key found, "" is entered into variable and then a 32 character check is performed
  api_key <- Sys.getenv('CLIMACELL_API')

  # ensure API key is 32 characters long (assumption that the API key is this long)
  # could not find confirmation of this
  if(nchar(api_key) < 32) {
    stop('API Key length is shorter than 32 characters, please recheck')
  }

  if(missing(lat)) {
    stop("No latitude provided")
  }

  if(missing(long)) {
    stop("No longitude provided")
  }

  if(is.numeric(lat) == FALSE | is.na(as.numeric(lat))) {
    stop("Latitude must be a numeric value or a string that can be coerced to numeric.")
  }

  if(is.numeric(long) == FALSE | is.na(as.numeric(long))) {
    stop("Longitude must be a numeric value or a string that can be coerced to numeric.")
  }

  if(missing(timestep)) {
    stop("No timestep provided")
  }

  # a timestep of current cannot have start or end times
  if((timestep == 'current' & !is.null(start_time)) | timestep == 'current' & !is.null(end_time)) {
    message("Timestep of 'current' cannot have start or end times.")
    start_time = NULL
    end_time = NULL
  }

  # timestep can only have certain values
  if(!timestep %in% c('1d', '1h', '30m','15m','5m','1m','current')) {
    stop("Timestep value is incorrect. \nAcceptable values are (choose one only): c('current','1m','5m','15m','30m','1h','1d')")
  }

  if(timestep != 'current' & is.null(start_time)) {
    message("No start time provided. Using current system time in UTC for start_time.")
    start_time <- lubridate::format_ISO8601(lubridate::with_tz(time = Sys.time(), tzone = 'UTC'), usetz = T)
  } else {
    # if user has already provided a start_time parameter value
    start_time <- lubridate::with_tz(time = start_time, tzone = 'UTC')
    start_time <- lubridate::format_ISO8601(x = start_time, usetz = T)
  }

  if(timestep %in% c('1m','5m', '15m','30m') & is.null(end_time)) {
    message("No end time provided. End time will be adjusted for 6 hours from current system time!")
    end_time <- lubridate::with_tz(time = Sys.time() + lubridate::dhours(6), tzone = 'UTC')
    end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
  } else if (timestep == '1h' & is.null(end_time)) {
    message("No end time provided. End time will be adjusted for 108 hours from current system time!")
    end_time <- lubridate::with_tz(time = Sys.time() + lubridate::dhours(108), tzone = 'UTC')
    end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
  } else if (timestep == '1d' & is.null(end_time)) {
    message("No end time provided. End time will be adjusted for 15 days from current system time!")
    end_time <- lubridate::with_tz(time = Sys.time() + lubridate::ddays(15), tzone = 'UTC')
    end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
  } else {
    end_time <- lubridate::with_tz(time = end_time, tzone = 'UTC')
    end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
  }

  # make sure start time is before end time
  if(start_time > end_time) {
    stop("Start time cannot be later than the end time!")
  }

  # error handling of the start & end times based on timestep used

  # timesteps 1m, 15m, 30m only allow 6 hours max into future from current actual time
  # timstep 1h allows for only 108 hours max into future from current actual time
  # timestep 1d allows for only 15 days max into future from current actual time

  # establish current time and the user times all in UTC
  current_time <- lubridate::with_tz(lubridate::now(), tzone = 'UTC')
  user_st_utc <- lubridate::ymd_hms(start_time)
  user_et_utc <- lubridate::ymd_hms(end_time)

  # define the maximum allowable time based on timestep
  if(timestep %in% c('1m','15m','30m')) {
    max_time_utc <- current_time + lubridate::dhours(6)
  } else if(timestep %in% c('1h')) {
    max_time_utc <- current_time + lubridate::dhours(108)
  } else if(timestep %in% c('1d')) {
    max_time_utc <- current_time + lubridate::ddays(15)
  }

  # if timestep is 1m, 15m, 30m, 1h then make sure start time is not 6 hours prior to current time

  # st_cur_chk: difference between the current system time and user start time
  # (< -6 means user start time is more than 6 hours prior to current system time)
  st_cur_chk <- lubridate::time_length(lubridate::interval(start = current_time, end = user_st_utc), unit = 'hours')

  if(st_cur_chk < -6.0) {
    message('Start time is more than 6 hours prior to the current system time!')
    message('Start time has been readjusted to be no more than 6 hours prior to current system time.')
    start_time <- lubridate::with_tz(lubridate::now() - lubridate::dhours(6), tzone = 'UTC')
    start_time <- lubridate::format_ISO8601(x = start_time, usetz = T)
  }


  # determine difference in time between the start time and the maximum allowable time
  # st_max_chk: difference between the maximum limit time and the user start time (if <1, <15, or < 30 (depending on timestep), there is a problem)
  st_max_chk <- lubridate::time_length(lubridate::interval(start = user_st_utc, end = max_time_utc), unit = 'minutes')

  if(timestep == '1m' & st_max_chk < 1.0) {
    stop("Start time cannot be less than 1 minute from maximum allowable time for this timestep!")
  } else if(timestep == '15m' & st_max_chk < 15.0) {
    stop("Start time cannot be less than 15 minutes from maximum allowable time for this timestep!")
  } else if(timestep == '30m' & st_max_chk < 30.0) {
    stop("Start time cannot be less than 30 minutes from maximum allowable time for this timestep!")
  } else if(timestep == '1h' & st_max_chk < 60.0) {
    stop("Start time cannot be less than 1 hour from maximum allowable time for this timestep!")
  } else if(timestep == '1d' & st_max_chk < 1440.0) {
    stop("Start time cannot be less than 24 hours from maximum allowable time for this timestep!")
  }

  # determine difference in time between end time and maximum allowable time
  # et_max_chk: difference between maximum limit time and user end time (if < 0, user end time is more than limit)
  et_max_chk <- lubridate::time_length(lubridate::interval(start = user_et_utc, end = max_time_utc), unit = 'hours')

  if(et_max_chk < 0) {
    if(timestep %in% c('1m','15m','30m')) {
      message('End time is more than 6 hours ahead of the current system time!')
      message('End time has been readjusted to be no more than 6 hours ahead of current system time.')
      end_time <- lubridate::with_tz(lubridate::now() + lubridate::dhours(6), tzone = 'UTC')
      end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
    } else if(timestep == '1h') {
      message('End time is more than 108 hours ahead of the current system time!')
      message('End time has been readjusted to be no more than 108 hours ahead of current system time.')
      end_time <- lubridate::with_tz(lubridate::now() + lubridate::dhours(108), tzone = 'UTC')
      end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
    } else if(timestep == '1d') {
      message('End time is more than 15 days ahead of the current system time!')
      message('End time has been readjusted to be no more than 15 days ahead of current system time.')
      end_time <- lubridate::with_tz(lubridate::now() + lubridate::dhours(15), tzone = 'UTC')
      end_time <- lubridate::format_ISO8601(x = end_time, usetz = T)
    }
  }

  # need to make sure there is adequate difference between start & end times provided by user

  et_st_chk <- lubridate::time_length(lubridate::interval(start = user_st_utc, end = user_et_utc), unit = 'minutes')

  if(timestep == '1m' & et_st_chk < 1.0) {
    stop('Difference between start time and end time cannot be less than 1 minute!')
  } else if(timestep == '15m' & et_st_chk < 15.0) {
    stop('Difference between start time and end time cannot be less than 15 minutes!')
  } else if(timestep == '30m' & et_st_chk < 30.0) {
    stop('Difference between start time and end time cannot be less than 30 minutes!')
  } else if(timestep == '1h' & et_st_chk < 60.0) {
    stop('Difference between start time and end time cannot be less than 60 minutes!')
  } else if(timestep == '1d' & et_st_chk < 1440.0) {
    stop('Difference between start time and end time cannot be less than 24 hours!')
  }

  # messsage for timestep other than 1D
  if(timestep != '1d') {
    message("Moonphase, Sunrise Time, and Sunset Times are only available if timestep is '1d'.")
  }

  # process for API retrieval

  latlong <- paste0(lat, ', ', long)

  # get results

  if(timestep != '1d') {
    result <- httr::content(
      httr::GET(
        url = 'https://data.climacell.co/v4/timelines',
        httr::add_headers('apikey'= api_key),
        httr::add_headers('content-type:' = 'application/json'),
        query = list(location = latlong,
                     fields = 'temperature',
                     fields = 'temperatureApparent',
                     fields = 'dewPoint',
                     fields = 'humidity',
                     fields = 'precipitationIntensity',
                     fields = 'precipitationProbability',
                     fields = 'precipitationType',
                     fields = 'windSpeed',
                     fields = 'windDirection',
                     fields = 'windGust',
                     fields = 'visibility',
                     fields = 'pressureSurfaceLevel',
                     fields = 'pressureSeaLevel',
                     fields = 'cloudCover',
                     fields = 'cloudBase',
                     fields = 'cloudCeiling',
                     fields = 'solarGHI',
                     fields = 'weatherCode',
                     timesteps = timestep,
                     startTime = start_time,
                     endTime = end_time
        )
      )
    )
  } else {
    result <- httr::content(
      httr::GET(
        url = 'https://data.climacell.co/v4/timelines',
        httr::add_headers('apikey'= api_key),
        httr::add_headers('content-type:' = 'application/json'),
        query = list(location = latlong,
                     fields = 'temperature',
                     fields = 'temperatureApparent',
                     fields = 'dewPoint',
                     fields = 'humidity',
                     fields = 'precipitationIntensity',
                     fields = 'precipitationProbability',
                     fields = 'precipitationType',
                     fields = 'windSpeed',
                     fields = 'windDirection',
                     fields = 'windGust',
                     fields = 'visibility',
                     fields = 'pressureSurfaceLevel',
                     fields = 'pressureSeaLevel',
                     fields = 'cloudCover',
                     fields = 'cloudBase',
                     fields = 'cloudCeiling',
                     fields = 'solarGHI',
                     fields = 'weatherCode',
                     fields = 'sunriseTime',
                     fields = 'sunsetTime',
                     fields = 'moonPhase',
                     timesteps = timestep,
                     startTime = start_time,
                     endTime = end_time
        )
      )
    )
  }

  tidy_result <- tibble::enframe(unlist(result))

  cln_result <- tidy_result %>%
    dplyr::mutate(
      name = gsub(pattern = 'data.timelines.', replacement = '', x = .data$name)
    ) %>%
    dplyr::filter(stringr::str_detect(pattern = 'intervals', string = .data$name))

  df_out <- cln_result %>%
    dplyr::mutate(
      start_time = ifelse(.data$name == 'intervals.startTime', .data$value, NA)
    ) %>%
    tidyr::fill(.data$start_time, .direction = 'down') %>%
    dplyr::filter(.data$name != 'intervals.startTime') %>%
    dplyr::mutate(
      var_name = gsub(pattern = 'intervals.values.', replacement = '', x = .data$name),
      var_name = dplyr::case_when(
        .data$var_name == 'temperature' ~ 'temp_c',
        .data$var_name == 'temperatureApparent' ~ 'temp_feel_c',
        .data$var_name == 'precipitationIntensity' ~ 'precipitation_intensity',
        .data$var_name == 'precipitationProbability' ~ 'precipitation_probability',
        .data$var_name == 'precipitationType' ~ 'precipitation_type_code',
        .data$var_name == 'dewPoint' ~ 'dewpoint',
        .data$var_name == 'windSpeed' ~ 'wind_speed',
        .data$var_name == 'windDirection' ~ 'wind_direction',
        .data$var_name == 'windGust' ~ 'wind_gust',
        .data$var_name == 'pressureSurfaceLevel' ~ 'pressure_surface_level',
        .data$var_name == 'pressureSeaLevel' ~ 'pressure_sea_level',
        .data$var_name == 'cloudCover' ~ 'cloud_cover',
        .data$var_name == 'solarGHI' ~ 'solar_ghi',
        .data$var_name == 'weatherCode' ~ 'weather_code',
        .data$var_name == 'cloudBase' ~ 'cloud_base',
        .data$var_name == 'cloudCeiling' ~ 'cloud_ceiling',
        .data$var_name == 'humidity' ~ 'humidity',
        .data$var_name == 'visibility' ~ 'visibility',
        .data$var_name == 'sunriseTime' ~ 'sunrise_time',
        .data$var_name == 'sunsetTime' ~ 'sunset_time',
        .data$var_name == 'moonPhase' ~ 'moon_phase_code'
      )
    ) %>%
    dplyr::filter(!is.na(.data$var_name)) %>%
    dplyr::select(-.data$name) %>%
    tidyr::pivot_wider(
      names_from = .data$var_name,
      values_from = .data$value
    )

  colchk_temp <- assertthat::has_name(df_out, 'temp_c')
  colchk_feel <- assertthat::has_name(df_out, 'temp_feel_c')
  colchk_precipint <- assertthat::has_name(df_out, 'precipitation_intensity')
  colchk_precipprob <- assertthat::has_name(df_out, 'precipitation_probability')
  colchk_preciptype <- assertthat::has_name(df_out, 'precipitation_type_code')
  colchk_dew <- assertthat::has_name(df_out, 'dewpoint')
  colchk_wind <- assertthat::has_name(df_out, 'wind_speed')
  colchk_winddir <- assertthat::has_name(df_out, 'wind_direction')
  colchk_windgust <- assertthat::has_name(df_out, 'wind_gust')
  colchk_prssurf <- assertthat::has_name(df_out, 'pressure_surface_level')
  colchk_prssea <- assertthat::has_name(df_out, 'pressure_sea_level')
  colchk_cldcvr <- assertthat::has_name(df_out, 'cloud_cover')
  colchk_solar <- assertthat::has_name(df_out, 'solar_ghi')
  colchk_weacode <- assertthat::has_name(df_out, 'weather_code')
  colchk_cldbse <- assertthat::has_name(df_out, 'cloud_base')
  colchk_cldceil <- assertthat::has_name(df_out, 'cloud_ceiling')
  colchk_humid <- assertthat::has_name(df_out, 'humidity')
  colchk_vis <- assertthat::has_name(df_out, 'visibility')
  colchk_moon <- assertthat::has_name(df_out, 'moon_phase_code')
  colchk_sunrise <- assertthat::has_name(df_out, 'sunrise_time')
  colchk_sunset <- assertthat::has_name(df_out, 'sunset_time')

  # if column does not exist, add

  if(colchk_temp == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(temp_c = NA)
  }
  if(colchk_feel == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(temp_feel_c = NA)
  }
  if(colchk_precipint == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(precipitation_intensity = NA)
  }
  if(colchk_precipprob == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(precipitation_probability = NA)
  }
  if(colchk_preciptype == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(precipitation_type_code = NA)
  }
  if(colchk_dew == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(dewpoint = NA)
  }
  if(colchk_wind == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(wind_speed = NA)
  }
  if(colchk_winddir == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(wind_direction = NA)
  }
  if(colchk_windgust == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(wind_gust = NA)
  }
  if(colchk_prssurf == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(pressure_surface_level = NA)
  }
  if(colchk_prssea == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(pressure_sea_level = NA)
  }
  if(colchk_cldcvr == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(cloud_cover = NA)
  }
  if(colchk_solar == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(solar_ghi = NA)
  }
  if(colchk_weacode == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(weather_code = NA)
  }
  if(colchk_cldbse == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(cloud_base = NA)
  }
  if(colchk_cldceil == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(cloud_ceiling = NA)
  }
  if(colchk_humid == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(humidity = NA)
  }
  if(colchk_vis == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(visibility = NA)
  }

  if(colchk_moon == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(moon_phase_code = NA)
  }
  if(colchk_sunrise == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(sunrise_time = NA)
  }
  if(colchk_sunset == FALSE) {
    df_out <- df_out %>%
      tibble::add_column(sunset_time = NA)
  }

  # encode each column correctly
  df_out_encoded <- df_out %>%
    dplyr::mutate(
      start_time = lubridate::ymd_hms(.data$start_time, tz = 'UTC'),
      sunrise_time = lubridate::ymd_hms(.data$sunrise_time, tz = 'UTC'),
      sunset_time = lubridate::ymd_hms(.data$sunset_time, tz = 'UTC')
    ) %>%
    dplyr::mutate(
      dplyr::across(where(is.character), .fns = as.numeric)
    )

  # add in the descriptions and rearrange
  df_final <- df_out_encoded %>%
    dplyr::left_join(precip_dict, by = 'precipitation_type_code') %>%
    dplyr::left_join(weather_dict, by = 'weather_code') %>%
    dplyr::select(.data$start_time,
                  .data$temp_c,
                  .data$temp_feel_c,
                  .data$weather_code,
                  .data$weather_desc,
                  .data$dewpoint,
                  .data$humidity,
                  .data$wind_speed,
                  .data$wind_direction,
                  .data$wind_gust,
                  .data$solar_ghi,
                  .data$precipitation_type_code,
                  .data$precipitation_type_desc,
                  .data$precipitation_probability,
                  .data$precipitation_intensity,
                  .data$cloud_cover,
                  .data$cloud_base,
                  .data$cloud_ceiling,
                  .data$visibility,
                  .data$pressure_surface_level,
                  .data$pressure_sea_level,
                  .data$sunrise_time,
                  .data$sunset_time,
                  .data$moon_phase_code
    )

  if(timestep == '1d') {
    df_final <- df_final %>%
      dplyr::left_join(moon_dict, by = 'moon_phase_code')
  } else {
    df_final <- df_final %>%
      dplyr::select(-.data$sunrise_time, -.data$sunset_time, -.data$moon_phase_code)
  }

  return(df_final)

}
